gnovm/pkg/doc/doc.go

```diff
@@ -0,0 +1,303 @@
+// Package doc implements support for documentation of Gno packages and realms,
+// in a similar fashion to `go doc`.
+// As a reference, the [official implementation] for `go doc` is used.
+//
+// [official implementation]: https://github.com/golang/go/tree/90dde5dec1126ddf2236730ec57511ced56a512d/src/cmd/doc
+package doc
+
+import (
+	"fmt"
+	"go/ast"
+	"go/doc"
+	"go/token"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"go.uber.org/multierr"
+)
+
+// WriteDocumentationOptions represents the possible options when requesting
+// documentation through Documentable.
+type WriteDocumentationOptions struct {
+	// ShowAll shows all symbols when displaying documentation about a package.
+	ShowAll bool
+	// Source shows the source code when documenting a symbol.
+	Source bool
+	// Unexported shows unexported symbols as well as exported.
+	Unexported bool
+	// Short shows a one-line representation for each symbol.
+	Short bool
+
+	w io.Writer
+}
+
+// Documentable is a package, symbol, or accessible which can be documented.
+type Documentable interface {
+	WriteDocumentation(w io.Writer, opts *WriteDocumentationOptions) error
+}
+
+// static implementation check
+var _ Documentable = (*documentable)(nil)
+
+type documentable struct {
+	Dir
+	symbol     string
+	accessible string
+	pkgData    *pkgData
+}
+
+func (d *documentable) WriteDocumentation(w io.Writer, o *WriteDocumentationOptions) error {
+	if o == nil {
+		o = &WriteDocumentationOptions{}
+	}
+	o.w = w
+
+	var err error
+	// pkgData may already be initialised if we already had to look to see
+	// if it had the symbol we wanted; otherwise initialise it now.
+	if d.pkgData == nil {
+		d.pkgData, err = newPkgData(d.Dir, o.Unexported)
+		if err != nil {
+			return err
+		}
+	}
+
+	astpkg, pkg, err := d.pkgData.docPackage(o)
+	if err != nil {
+		return err
+	}
+
+	// copied from go source - map vars, constants and constructors to their respective types.
+	typedValue := make(map[*doc.Value]bool)
+	constructor := make(map[*doc.Func]bool)
+	for _, typ := range pkg.Types {
+		pkg.Consts = append(pkg.Consts, typ.Consts...)
+		pkg.Vars = append(pkg.Vars, typ.Vars...)
+		pkg.Funcs = append(pkg.Funcs, typ.Funcs...)
+		if o.Unexported || token.IsExported(typ.Name) {
+			for _, value := range typ.Consts {
+				typedValue[value] = true
+			}
+			for _, value := range typ.Vars {
+				typedValue[value] = true
+			}
+			for _, fun := range typ.Funcs {
+				// We don't count it as a constructor bound to the type
+				// if the type itself is not exported.
+				constructor[fun] = true
+			}
+		}
```

Comment:
good catch, although because of the name that I bear it is my due obligation to be annoying and point out that [de Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) state that the correct negation in this case is `!o.Unexported && !token.IsExported(typ.Name)` :wink: :stuck_out_tongue_closed_eyes:

Reply:
ahaha, totally, I forgot to completely negate the expression changing the `||` to `&&`, thanks for pointing it out.

Reply:
true, sorry about that! sometimes is difficult to head-compile meanwhile doing morning code reviews :).